### PR TITLE
Fix DP attention on CPU

### PR DESCRIPTION
**Pull request by chunyuan-w**
## Motivation

Fix the functionality of DP attention on CPU.

## Modifications

1. For distributed OPs, use `torch.distributed.all_gather_into_tensor` and `torch.distributed.reduce_scatter_tensor` on CPU for now to fix the functionality. We will submit follow-up PRs to add optimized kernels.
2. Fix the `tp_size` used in `intel_amx_backend`.
3. Added a reference CPU implementation of `memcpy` as a counterpart to `memcpy_triton` in `dp_attention`. We will submit follow-up PRs to add optimized kernels if needed.
4. For MoE CPU, whether all_reduce is needed after the MoE kernel is incorrect when dp is on (`use_reduce_scatter` is not checked in the current code). Fixed it by aligning with GPU:
https://github.com/sgl-project/sglang/blob/90401cf7d243da5dcb4f2cef527e2b342ea468b5/python/sglang/srt/models/deepseek_v2.py#L859-L865
5. Added a UT for DP attention on `DeepSeek-Coder-V2-Lite-Instruct`.
## Accuracy Tests

Model: `deepseek-ai/DeepSeek-Coder-V2-Lite-Instruct`

<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/chunyuan/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/chunyuan/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">

</head>

<body link="#0563C1" vlink="#954F72">

  | device | mmlu | gsm8k
-- | -- | -- | --
DP=2, TP=2 | cpu| 0.604 | 0.805
DP=2, TP=2 | cuda | 0.605 | 0.8

</body>

</html>

```sh
# Cmd for CPU server
export SGLANG_USE_CPU_ENGINE=1
python3 -m sglang.launch_server --model deepseek-ai/DeepSeek-Coder-V2-Lite-Instruct --disable-radix-cache --trust-remote-code --device cpu --log-requests --log-requests-level 3 --disable-overlap-schedule  --mem-fraction-static 0.05 --tp 2 --enable-dp-attention  --dp-size 2

# Cmd for GPU server
 python3 -m sglang.launch_server --model deepseek-ai/DeepSeek-Coder-V2-Lite-Instruct --disable-radix-cache --trust-remote-code --log-requests --log-requests-level 3 --disable-overlap-schedule --tp 2 --enable-dp-attention  --dp-size 2

# mmlu
cd benchmark/mmlu/
bash download_data.sh
python bench_sglang.py

# gsm8k
cd benchmark/gsm8k/
python bench_sglang.py
```

---
### CI States

Latest PR Test (Base): :no_entry_sign: [Run #32094328854](https://github.com/sgl-project/sglang/actions/runs/32094328854)
Latest PR Test (Extra): :x: [Run #32094328478](https://github.com/sgl-project/sglang/actions/runs/32094328478)

### CODE REVIEW

**Comment by chunyuan-w** [comment]
@mingfeima could you please help review this PR?

**Comment by mingfeima** [comment]
need to rebase to fix the Xeon CI failures.

File: python/sglang/srt/layers/attention/intel_amx_backend.py [hunk 1]
```diff
@@ -21,7 +22,7 @@ def __init__(self, model_runner: ModelRunner):
         self.device = model_runner.device
 
         self.num_head = (
-            model_runner.model_config.num_attention_heads // model_runner.tp_size
+            model_runner.model_config.num_attention_heads // get_attention_tp_size()
```
Comment (mingfeima) [inline python/sglang/srt/layers/attention/intel_amx_backend.py]: what's the difference of `get_attention_tp_size` with `model_runner.tp_size`

File: python/sglang/srt/layers/dp_attention.py [hunk 2]
```diff
@@ -425,6 +426,33 @@ def prod(x):
...
+        dst_start = offset
+
+    dst_end = min(dst_start + sz, total_rows_dst)
+    src_end = min(src_start + sz, total_rows_src)
+    actual_sz = min(dst_end - dst_start, src_end - src_start)
+
+    if actual_sz <= 0:
+        return
+
+    dst[dst_start : dst_start + actual_sz].copy_(src[src_start : src_start + actual_sz])
```
Comment (mingfeima) [inline python/sglang/srt/layers/dp_attention.py:479]: this should be fairly easy to be implemented with avx512, replace it with sgl-kernel later on.

File: python/sglang/srt/layers/dp_attention.py [hunk 3]
```diff
@@ -436,6 +464,13 @@ def memcpy_triton(dst, src, dim, offset, sz, offset_src):
     memcpy_triton_kernel[grid](dst, src, offset, sz, offset_src, chunk_size, BLOCK_SIZE)
 
 
+def memcpy(dst, src, dim, offset, sz, offset_src):
+    if _is_cpu:
+        memcpy_cpu(dst, src, dim, offset, sz, offset_src)
+    else:
+        memcpy_triton(dst, src, dim, offset, sz, offset_src)
```
Comment (mingfeima) [inline python/sglang/srt/layers/dp_attention.py]:
```suggestion
def memcpy(dst, src, dim, offset, sz, offset_src):
    func = memcpy_cpu if _is_cpu else memcpy_triton
    func(dst, src, dim, offset, sz, offset_src)
```

Comment (mingfeima) [reply python/sglang/srt/layers/dp_attention.py]: actully make this `func` global would be better, save python if-else.

File: python/sglang/srt/model_executor/forward_batch_info.py [hunk 4]
```diff
@@ -772,16 +773,24 @@ def prepare_mlp_sync_batch(self, model_runner: ModelRunner):
...
         self._pad_inputs_to_size(model_runner, num_tokens, bs)
         self.global_num_tokens_cpu = global_num_tokens
-        global_num_tokens_pinned = torch.tensor(global_num_tokens, pin_memory=True)
-        self.global_num_tokens_gpu.copy_(global_num_tokens_pinned, non_blocking=True)
+        global_num_tokens_pinned = torch.tensor(
+            global_num_tokens, pin_memory=not _is_cpu
+        )
+        self.global_num_tokens_gpu.copy_(
+            global_num_tokens_pinned, non_blocking=not _is_cpu
+        )
```
Comment (mingfeima) [inline python/sglang/srt/model_executor/forward_batch_info.py:1468]: reduce python if-else as much as possible:
```
self.use_pin_memory = not _is_cpu
```

**Review by mingfeima — CHANGES_REQUESTED** [review]
generally LGTM, just some minor changes needed.

let's provide sgl-kernels as well and evaluate the performance internally first!

File: python/sglang/srt/layers/attention/intel_amx_backend.py (hunk 1 above)
Comment (chunyuan-w) [reply python/sglang/srt/layers/attention/intel_amx_backend.py]: When dp attention is **disabled**, `attn_tp_size = tp_size`.
When dp attention is **enabled**,  `attn_tp_size = tp_size // dp_size`.

`get_attention_tp_size()` will call the below function to calculate `attn_tp_size` both both the case dp attention is enabled or not. 
https://github.com/sgl-project/sglang/blob/0c966779022a52224d4d4883d42accd7230ee1f2/python/sglang/srt/layers/dp_attention.py#L227-L235


Other attention backend also uses `get_attention_tp_size()`: 
https://github.com/sgl-project/sglang/blob/0c966779022a52224d4d4883d42accd7230ee1f2/python/sglang/srt/layers/attention/triton_backend.py#L89-L94

File: python/sglang/srt/model_executor/forward_batch_info.py (hunk 4 above)
Comment (chunyuan-w) [reply python/sglang/srt/model_executor/forward_batch_info.py:1468]: Fixed in https://github.com/sgl-project/sglang/pull/12961/commits/23b3927dab228a6f6d068d647d5ef697ef459816

File: python/sglang/srt/layers/dp_attention.py (hunk 3 above)
Comment (chunyuan-w) [reply python/sglang/srt/layers/dp_attention.py]: Changed the func to be global in https://github.com/sgl-project/sglang/pull/12961/commits/23b3927dab228a6f6d068d647d5ef697ef459816

**Review by mingfeima — APPROVED** [review]

File: python/sglang/srt/models/deepseek_v2.py [hunk 5]
```diff
@@ -1193,7 +1193,11 @@ def forward_cpu(
             ),  # block_size
             True,  # is_vnni
         )
-        if self.tp_size > 1 and not get_forward().fuse_mlp_allreduce:
+        if (
+            self.tp_size > 1
+            and not get_forward().fuse_mlp_allreduce
+            and not get_forward().mlp_reduce_scatter
+        ):
```
Comment (mingfeima) [inline python/sglang/srt/models/deepseek_v2.py]: https://github.com/sgl-project/sglang/pull/12961/changes#diff-5b9e34dd492bd8a14702a18b594721091092276fad1cf8736fba6ef1f33c1b04R1135-R1137

you can directly use `should_skip_post_experts_all_reduce` as cuda in L1135

File: python/sglang/srt/model_executor/forward_batch_info.py [hunk 6]
```diff
@@ -1363,8 +1370,12 @@ def _pad_inputs_to_size(self, model_runner: ModelRunner, num_tokens, bs):
...
             self.lora_ids.extend((bs - len(self.lora_ids)) * [None])
 
+        from sglang.srt.layers.attention.intel_amx_backend import IntelAMXAttnBackend
+
         seq_len_fill_value = (
-            model_runner.attn_backend.get_cuda_graph_seq_len_fill_value()
+            model_runner.attn_backend.get_cpu_graph_seq_len_fill_value()
+            if isinstance(model_runner.attn_backend, IntelAMXAttnBackend)
+            else model_runner.attn_backend.get_cuda_graph_seq_len_fill_value()
         )
```
Comment (mingfeima) [inline python/sglang/srt/model_executor/forward_batch_info.py]: intel_amx_backend.py depends on `ForwardBatch`：

in intel_amx_backend.py
`from sglang.srt.model_executor.forward_batch_info import ForwardBatch`

it will be a loop:
```
forward_batch_info  import IntelAMXAttnBackend
intel_amx_backend  import  ForwardBatch
```

here you are trying to tell whether this backend has `get_cpu_graph_seq_len_fill_value`

you can do something like:
```
if hasattr(model_runner.attn_backend, "get_cpu_graph_seq_len_fill_value"):
    seq_len_fill_value = model_runner.attn_backend.get_cpu_graph_seq_len_fill_value()
else:
    seq_len_fill_value = model_runner.attn_backend.get_cuda_graph_seq_len_fill_value()
```

**Review by mingfeima — CHANGES_REQUESTED** [review]
some minor changes are needed.

File: python/sglang/srt/models/deepseek_v2.py (hunk 5 above)
Comment (chunyuan-w) [reply python/sglang/srt/models/deepseek_v2.py]: Fixed

File: python/sglang/srt/model_executor/forward_batch_info.py (hunk 6 above)
Comment (chunyuan-w) [reply python/sglang/srt/model_executor/forward_batch_info.py]: I found that using the above `hasattr` way may introduce issues for cuda:

`HybridLinearAttnBackend` unconditionally defines both `get_cpu_graph_seq_len_fill_value` and `get_cuda_graph_seq_len_fill_value` https://github.com/sgl-project/sglang/blob/bd3f6a79358cd6f21639b8bd2ea03dd4f90bd3bf/python/sglang/srt/layers/attention/hybrid_linear_attn_backend.py#L916-L920

So fo a hybrid model running on GPU, where `full_attn_backend` is a plain CUDA backend, `hasattr(attn_backend, "get_cpu_graph_seq_len_fill_value")` is still `True`, and we will call `full_attn_backend.get_cpu_graph_seq_len_fill_value()` on a CUDA backend, which is incorrect.

I updated the code to use `_is_cpu` to decide whether `.get_cpu_graph_seq_len_fill_value` should be used. This avoids the circular import and won't have the above issue on cuda. Could you please check if this is okay?

**Review by mingfeima — APPROVED** [review]

**Comment by mingfeima** [comment]
@chunyuan-w update cookbook if necessary.

**Comment by chunyuan-w** [comment]
> @chunyuan-w update cookbook if necessary.

Sure, well noted.
